platform-sdk/docs/proposals/001-tss-rosters-design-proposal/001-tss-rosters-design-proposal.md

```diff
@@ -0,0 +1,265 @@
+consensus and block signing.
+Rosters, as immutable subsets of the address book, will provide this mechanism, ensuring efficient and secure key
+management for TSS operations.
+This proposal attempts to provide a specification for the behavior of rosters starting from their creation from the CAB
+to their terminal state within the TSS specification.
+
+### Requirements
+
+- Immutability: Rosters must be immutable to ensure the integrity of the computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and avoid unnecessary
+  data copying.
+- Clear API: A well-defined API should be provided for the Hedera app to create and submit candidate rosters, and for
+  the platform to manage the lifecycle of submitted rosters.
+- Reconnect Compatibility: The reconnect process must be adapted to use rosters instead of address books for validation
+  and node selection.
+- Address book paradigm discontinued within the platform codebase, only use rosters instead
+
+### Changes
+
+### Architecture and/or Components
+
+###### Roster API
+
+The Hedera app is already responsible for managing the address book. We propose that it continues to do so, and at
+adoption time, creates a candidate Roster object from the CAB and set it in the state. The State will contain one map of
+rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```. It should be noteworthy that the roster must be
+immutable so as to guarantee the integrity of the computed hash. This map of rosters will typically contain the
+currently active roster, the optional previously active roster, and an optional candidate roster. There will be new
+fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time, the way to trigger
+the adoption of a roster will be by the client code inserting a roster in the roster map, alongside setting the
+candidateRosterHash field in the PlatformState. The `activeRosterHash` will be private to the platform and will not be
+settable or modifiable from outside the platform.
+
+This indirection avoids moving data around in the merkle state which requires data to be copied into the block stream.
+This is computationally expensive. Another benefit of this approach is that adoption trigger becomes very simple (app
+sets the roster) with clear responsibility between the app and the platform.
+
+Some edge cases worth considering include:
+
+1. multiple concurrent roster submission - has a potential for introducing race conditions. We will prevent this by
+   guaranteeing immutability on rosters.
+2. Size control on the map of rosters - we certainly don’t want this map to grow infinitely so insertion of rosters will
+   be controlled, with clear rules for removal of unused rosters (e.g. the acceptance of a new candidate roster
+   invalidates and removes existing one).
```

Comment:
```suggestion
   be controlled, with clear rules for removal of unused rosters (e.g. the acceptance of a new candidate roster
   invalidates and removes the previous roster, and marks the currently active roster as the new previous roster).
```

And while the above is good to handle the previous/active/candidate rosters, there might be a case where multiple candidate rosters may still end up in the map if some of them never get adopted.

I suggest to consider introducing a TTL for rosters and have a periodic process that would scan the map and delete all entries past their TTL unless they're currently previous/active/candidate roster. If the TTL is implemented, then on average the map size will never exceed 3.

There's a caveat related to that: should the app choose to use the map to hosts their future rosters, it would have to be aware of this clean up procedure. Alternatively, we can explicitly prohibit the app to host future rosters in this map.